test/ssc_test/cmod_battery_pvsamv1_test.cpp

```diff
@@ -0,0 +1,448 @@
+
+			auto batt_power = data_vtab->as_vector_ssc_number_t("batt_power");
+			daily_battery_stats batt_stats = daily_battery_stats(batt_power);
+
+			EXPECT_NEAR(batt_stats.peakKwCharge, peakKwCharge[i], m_error_tolerance_lo);
+			EXPECT_NEAR(batt_stats.peakKwDischarge, peakKwDischarge[i], m_error_tolerance_lo);
+			EXPECT_NEAR(batt_stats.peakCycles, peakCycles[i], m_error_tolerance_lo);
+			EXPECT_NEAR(batt_stats.avgCycles, avgCycles[i], 0.0001); // Runs once per year
+		}
+	}
+}
+
+/// Test PVSAMv1 with all defaults and battery enabled with custom dispatch
+TEST_F(CMPvsamv1BatteryIntegration_cmod_pvsamv1, ResidentialACDCBatteryModelIntegrationCustomDispatch)
+{
+	ssc_data_t data = ssc_data_create();
+	pvsamv_nofinancial_default(data);
+	battery_data_default(data);
+
+	std::map<std::string, double> pairs;
+	pairs["en_batt"] = 1;
+
+	pairs["analysis_period"] = 1;
+	set_array(data, "load", load_profile_path, 8760); // Load is required for peak shaving controllers
+	pairs["batt_dispatch_choice"] = 3;
+	set_array(data, "batt_custom_dispatch", custom_dispatch_residential_schedule, 8760);
+
+	ssc_number_t expectedEnergy[2] = { 8710, 8717 };
+	ssc_number_t expectedBatteryChargeEnergy[2] = { 4.6, 4.7};
+	ssc_number_t expectedBatteryDischargeEnergy[2] = { 0.76, 7.6};
+
+	ssc_number_t peakKwCharge[2] = { -2.7, -2.5};
+	ssc_number_t peakKwDischarge[2] = { 0.6, 4.0};
+	ssc_number_t peakCycles[2] = { 1, 1 };
+	ssc_number_t avgCycles[2] = { 0.0027, 0.0027};
+
+	// Test both AC and DC using the same dispatch model
+	for (int i = 0; i < 2; i++) {
+		pairs["batt_ac_or_dc"] = i;
+
+		int pvsam_errors = modify_ssc_data_and_run_module(data, "pvsamv1", pairs);
+		EXPECT_FALSE(pvsam_errors);
+
+		if (!pvsam_errors)
+		{
+			ssc_number_t annual_energy;
+			ssc_data_get_number(data, "annual_energy", &annual_energy);
+			EXPECT_NEAR(annual_energy, expectedEnergy[i], m_error_tolerance_hi) << "Annual energy.";
+
+			auto data_vtab = static_cast<var_table*>(data);
+			auto annualChargeEnergy = data_vtab->as_vector_ssc_number_t("batt_annual_charge_energy");
+			EXPECT_NEAR(annualChargeEnergy[0], expectedBatteryChargeEnergy[i], m_error_tolerance_hi) << "Battery annual charge energy.";
+
+			auto annualDischargeEnergy = data_vtab->as_vector_ssc_number_t("batt_annual_discharge_energy");
+			EXPECT_NEAR(annualDischargeEnergy[0], expectedBatteryDischargeEnergy[i], m_error_tolerance_hi) << "Battery annual discharge energy.";
+
+			auto batt_power = data_vtab->as_vector_ssc_number_t("batt_power");
+			daily_battery_stats batt_stats = daily_battery_stats(batt_power);
+
+			EXPECT_NEAR(batt_stats.peakKwCharge, peakKwCharge[i], m_error_tolerance_lo);
+			EXPECT_NEAR(batt_stats.peakKwDischarge, peakKwDischarge[i], m_error_tolerance_lo);
+			EXPECT_NEAR(batt_stats.peakCycles, peakCycles[i], m_error_tolerance_lo);
+			EXPECT_NEAR(batt_stats.avgCycles, avgCycles[i], 0.0001); // Runs once per year
```

Comment:
I really like this. I was wondering if you could compute `daily_battery_stats` on `custom_dispatch_residential_schedule` and see how the stats differ? It probably won't be very much due to the low activity.

Here's another dispatch csv you can try instead. I pulled it from the reopt sizing function for pvsamv1 residential a while back, and while it's not fitted to the parameters here, I think it's about the right battery size for the load.
[custom_dispatch_residential_hourly.csv.zip](https://github.com/NREL/ssc/files/4383903/custom_dispatch_residential_hourly.csv.zip)

Reply:
I like these sparse schedules as tests, since they give me insight into the battery stats over a longer period, and dispatch the system at capacity. I added your files to new tests, thanks for sending them!